File: 0003_max_depth_binary_tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }

    var l, r int

    if root.Left != nil {
        l = maxDepth(root.Left)
    }

    if root.Right != nil {
        r = maxDepth(root.Right)
    }

    if l == 0 && r == 0 {
        return 1
    }

    if l > r {
        return 1 + l
    }

    return 1 + r
}

File: 0003_max_depth_binary_tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;

        int leftMax = root.left == null ? 0 : maxDepth(root.left);
        int rightMax = root.right == null ? 0 : maxDepth(root.right);

        return 1 + Math.max(leftMax, rightMax);
    }
}
